src/tree/mod.rs

```diff
@@ -1,103 +1,363 @@
-        tree_history_size: usize,
-        address: H160,
-        creation_block: u64,
-        window_size: u64,
-        middleware: Arc<M>,
+        tree_depth: usize,
+        canonical_tree_manager: TreeManager<M, CanonicalTree>,
+        bridged_tree_manager: Vec<TreeManager<M, BridgedTree>>,
     ) -> Self {
+        let identity_tree = IdentityTree::new(tree_depth);
+
         Self {
-            tree_data: Arc::new(RwLock::new(TreeData::new(
-                tree,
-                tree_history_size,
-            ))),
-            tree_updater: Arc::new(TreeUpdater::new(
-                address,
-                creation_block,
-                window_size,
-                middleware,
-            )),
-            synced: Arc::new(AtomicBool::new(false)),
+            identity_tree: Arc::new(RwLock::new(identity_tree)),
+            canonical_tree_manager,
+            bridged_tree_manager,
+            chain_state: Arc::new(RwLock::new(HashMap::new())),
+            synced: AtomicBool::new(false),
         }
     }
 
-    /// Spawns a task that continually syncs the `TreeData` to the state at the chain head.
-    #[instrument(skip(self))]
-    pub fn spawn(&self) -> JoinHandle<Result<(), TreeAvailabilityError<M>>> {
-        let tree_data = self.tree_data.clone();
-        let tree_updater = self.tree_updater.clone();
+    pub async fn spawn(&self) -> eyre::Result<Vec<JoinHandle<()>>> {
+        let start_time = Instant::now();
 
-        tracing::info!("Spawning thread to sync tree");
-        let synced = self.synced.clone();
+        tracing::info!("Syncing to head");
+        self.sync_to_head().await?;
+        tracing::info!("Synced to head in {:?} seconds", start_time.elapsed());
 
-        tokio::spawn(async move {
-            let start = tokio::time::Instant::now();
-            tree_updater.sync_to_head(&tree_data).await?;
-            let sync_time = start.elapsed();
+        let (leaf_updates_tx, leaf_updates_rx) =
+            tokio::sync::mpsc::channel(100);
+        let (bridged_root_tx, bridged_root_rx) =
+            tokio::sync::mpsc::channel(100);
+
+        // Spawn the tree managers for the canonical and bridged trees
+        let mut handles = vec![];
+        handles.push(self.canonical_tree_manager.spawn(leaf_updates_tx));
+
+        //TODO: handle if there are no bridged roots to spawn, otherwise the channel will close
+        for bridged_tree in self.bridged_tree_manager.iter() {
+            handles.push(bridged_tree.spawn(bridged_root_tx.clone()));
+        }
+
+        handles
+            .push(self.handle_tree_updates(leaf_updates_rx, bridged_root_rx));
+
+        Ok(handles)
+    }
 
-            tracing::info!(?sync_time, "WorldTree synced to chain head");
-            synced.store(true, Ordering::Relaxed);
+    fn handle_tree_updates(
+        &self,
+        mut leaf_updates_rx: Receiver<(Root, LeafUpdates)>,
+        mut bridged_root_rx: Receiver<(u64, Hash)>,
```

Comment:
Consider a newtype `ChainId(pub u64)`

Reply:
Maybe something like this https://github.com/worldcoin/contract-deployer/blob/main/src/types.rs